File: 2015/kt/src/test/kotlin/org/adventofcode/Day08.kt
package org.adventofcode

class Day08() {
  val diff = { str: String ->
    val real = str
      .replace(Regex("""^""""), "")
      .replace(Regex(""""$"""), "")
      .replace(Regex("""\\[\"\\]"""), " ")
      .replace(Regex("""\\x.."""), " ")
    str.length - real.length
  }

  val diff2 = { str: String ->
    val real = str
      .replace(Regex("""[\\\"]"""), "  ")
    real.length - str.length + 2
  }

  private fun part(list: List<String>, fn: (String) -> Int): Int {
    return list.map(fn).sum()
  }

  fun part1(str: List<String>) = part(str, diff)
  fun part2(str: List<String>) = part(str, diff2)

}

File: 2015/js/day08.js
import '../../utils/js/array'

const memStr = str => str
  .replace(/^\"/g, "")
  .replace(/\"$/g, "")
  .replace(/\\[\\\"]/g, " ")
  .replace(/\\x../g, " ")


const diff = str => str.length - memStr(str).length
const part1 = list => list
  .map(diff)
  .sum()

const encodeStr = str => str
  .replace(/[\\\"]/g, "  ")

const diff2 = str => encodeStr(str).length - str.length + 2
const part2 = list => list
  .map(diff2)
  .sum()

export { diff, diff2, part1, part2 }